drivers/pwm/pwm_max31790.c

```diff
@@ -0,0 +1,677 @@
+	case 5:
+		MAX37190_PWMFREQUENCY_PWM4TO6_SET(value_pwm_frequency, pwm_frequency_channel);
+		break;
+	default:
+		LOG_ERR("invalid channel %i", channel);
+		return -EINVAL;
+	}
+
+	result = max31790_write_register(dev, MAX37190_PWMFREQUENCY, value_pwm_frequency);
+	if (result != 0) {
+		return result;
+	}
+
+	value_fan_configuration = 0;
+	value_fan_dynamics = 0;
+
+	if (flags & PWM_MAX31790_FLAG_SPIN_UP) {
+		MAX37190_FANXCONFIGURATION_SPINUP_SET(value_fan_configuration, 0b10);
+	} else {
+		MAX37190_FANXCONFIGURATION_SPINUP_SET(value_fan_configuration, 0b00);
+	}
+
+	MAX37190_FANXCONFIGURATION_MONITOR_SET(value_fan_configuration, 0b0);
+	MAX37190_FANXCONFIGURATION_LOCKEDROTOR_SET(value_fan_configuration, 0b0);
+	MAX37190_FANXCONFIGURATION_LOCKEDROTORPOLARITY_SET(value_fan_configuration, 0b0);
+	MAX37190_FANXCONFIGURATION_TACH_SET(value_fan_configuration, 0b0);
+	MAX37190_FANXCONFIGURATION_TACHINPUTENABLED_SET(value_fan_configuration, 0b1);
+
+	MAX37190_FANXDYNAMICS_SPEEDRANGE_SET(value_fan_dynamics, value_speed_range);
+	MAX37190_FANXDYNAMICS_PWMRATEOFCHANGE_SET(value_fan_dynamics, value_pwm_rate_of_change);
+	MAX37190_FANXDYNAMICS_ASYMMETRICRATEOFCHANGE_SET(value_fan_dynamics, 0b1);
+
+	switch (channel) {
+	case 0:
+		register_fan_configuration = MAX37190_FAN1CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT1TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT1TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN1DYNAMICS;
+		break;
+	case 1:
+		register_fan_configuration = MAX37190_FAN2CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT2TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT2TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN2DYNAMICS;
+		break;
+	case 2:
+		register_fan_configuration = MAX37190_FAN3CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT3TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT3TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN3DYNAMICS;
+		break;
+	case 3:
+		register_fan_configuration = MAX37190_FAN4CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT4TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT4TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN4DYNAMICS;
+		break;
+	case 4:
+		register_fan_configuration = MAX37190_FAN5CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT5TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT5TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN5DYNAMICS;
+		break;
+	case 5:
+		register_fan_configuration = MAX37190_FAN6CONFIGURATION;
+		register_pwm_target_duty_cycle_msb = MAX37190_PWMOUT6TARGETDUTYCYCLEMSB;
+		register_pwm_target_duty_cycle_lsb = MAX37190_PWMOUT6TARGETDUTYCYCLELSB;
+		register_fan_dynamics = MAX37190_FAN6DYNAMICS;
+		break;
+	default:
+		LOG_ERR("invalid channel %i", channel);
+		return -EINVAL;
+	}
```

Comment:
Would not you be better off validating "channel" and then computing the register offset?

Reply:
I would consider this a question of taste? I personally prefer to have these offsets explicitly set, makes it in my opinion easier to understand.

Reply:
Sure but it's 40 lines of code to implement an addition with a switch lookup table and the input value is already validated by the caller. I could understand if there was any advantage (readability? performance? code compactness? stack usage?) but I can't see any. This is how it's done in Linux for the same device: https://lxr.missinglinkelectronics.com/linux/drivers/hwmon/max31790.c#L19, other drivers in Zephyr use the same style for defining registers (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pwm/pwm_gd32.c#L50-L55).

Reply:
Will do